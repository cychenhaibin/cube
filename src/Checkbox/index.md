# Checkbox 复选框

复选框组件。

## 代码演示

### 基础用法

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => <Checkbox>Checkbox</Checkbox>;
```

### 受控组件

```jsx
import React from 'react';
import { Checkbox } from 'cube';
import { useState } from 'react';

export default () => {
  const [checked, setChecked] = useState(false);

  return (
    <Checkbox checked={checked} onChange={(checked) => setChecked(checked)}>
      Checkbox
    </Checkbox>
  );
};
```

### 禁用状态

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => (
  <>
    <Checkbox disabled>Disabled</Checkbox>
    <Checkbox disabled checked>
      Disabled Checked
    </Checkbox>
  </>
);
```

### 只读状态

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => (
  <>
    <Checkbox readOnly>ReadOnly</Checkbox>
    <Checkbox readOnly checked>
      ReadOnly Checked
    </Checkbox>
  </>
);
```

### 不同尺寸

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => (
  <>
    <Checkbox size="small">Small</Checkbox>
    <Checkbox>Default</Checkbox>
    <Checkbox size="large" checked>
      Large
    </Checkbox>
  </>
);
```

### 复选框组

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => (
  <Checkbox.Group>
    <Checkbox value="1" checked>
      Option 1
    </Checkbox>
    <Checkbox value="2">Option 2</Checkbox>
    <Checkbox value="3">Option 3</Checkbox>
  </Checkbox.Group>
);
```

### 垂直排列

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => (
  <Checkbox.Group direction="vertical">
    <Checkbox value="1" checked>
      Option 1
    </Checkbox>
    <Checkbox value="2">Option 2</Checkbox>
    <Checkbox value="3">Option 3</Checkbox>
  </Checkbox.Group>
);
```

### 使用 options

```jsx
import React from 'react';
import { Checkbox } from 'cube';

export default () => {
  const options = [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
  ];

  return <Checkbox.Group options={options} />;
};
```

## API

### Checkbox

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| checked | 指定当前是否选中 | `boolean` | - |
| defaultChecked | 初始是否选中 | `boolean` | - |
| disabled | 是否禁用 | `boolean` | false |
| readOnly | 是否只读 | `boolean` | false |
| value | 复选框的值 | `string` \| `number` | - |
| onChange | 变化时回调函数 | `(checked: boolean, value?: string \| number) => void` | - |
| size | 尺寸 | `small` \| `default` \| `large` | default |

### Checkbox.Group

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| value | 指定选中的选项 | `(string \| number)[]` | - |
| defaultValue | 默认选中的选项 | `(string \| number)[]` | - |
| options | 指定可选项 | `{ label: string; value: string \| number; disabled?: boolean }[]` | - |
| disabled | 是否禁用 | `boolean` | false |
| readOnly | 是否只读 | `boolean` | false |
| direction | 排列方向 | `horizontal` \| `vertical` | horizontal |
| onChange | 变化时回调函数 | `(values: (string \| number)[]) => void` | - |
| size | 尺寸 | `small` \| `default` \| `large` | default |
