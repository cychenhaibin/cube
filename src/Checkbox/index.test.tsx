import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Checkbox from './index';

describe('Checkbox', () => {
  it('renders correctly', () => {
    const { container } = render(<Checkbox>Checkbox</Checkbox>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be checked when checked prop is true', () => {
    render(<Checkbox checked>Checkbox</Checkbox>);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeChecked();
  });

  it('should be disabled when disabled prop is true', () => {
    render(<Checkbox disabled>Checkbox</Checkbox>);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeDisabled();
  });

  it('should be readonly when readOnly prop is true', () => {
    render(<Checkbox readOnly>Checkbox</Checkbox>);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveAttribute('readOnly');
  });

  it('should call onChange when clicked', () => {
    const onChange = jest.fn();
    render(<Checkbox onChange={onChange}>Checkbox</Checkbox>);
    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);
    expect(onChange).toHaveBeenCalledWith(true, undefined);
  });

  it('should not call onChange when disabled', () => {
    const onChange = jest.fn();
    render(
      <Checkbox disabled onChange={onChange}>
        Checkbox
      </Checkbox>,
    );
    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should not call onChange when readonly', () => {
    const onChange = jest.fn();
    render(
      <Checkbox readOnly onChange={onChange}>
        Checkbox
      </Checkbox>,
    );
    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);
    expect(onChange).not.toHaveBeenCalled();
  });
});

describe('Checkbox.Group', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Checkbox.Group>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders options correctly', () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' },
    ];
    render(<Checkbox.Group options={options} />);
    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();
  });

  it('should be disabled when disabled prop is true', () => {
    render(
      <Checkbox.Group disabled>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    const checkboxes = screen.getAllByRole('checkbox');
    checkboxes.forEach((checkbox) => {
      expect(checkbox).toBeDisabled();
    });
  });

  it('should be readonly when readOnly prop is true', () => {
    render(
      <Checkbox.Group readOnly>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    const checkboxes = screen.getAllByRole('checkbox');
    checkboxes.forEach((checkbox) => {
      expect(checkbox).toHaveAttribute('readOnly');
    });
  });

  it('should call onChange when a checkbox is clicked', () => {
    const onChange = jest.fn();
    render(
      <Checkbox.Group onChange={onChange}>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    const checkbox = screen.getByLabelText('Option 1');
    fireEvent.click(checkbox);
    expect(onChange).toHaveBeenCalledWith(['1']);
  });

  it('should not call onChange when disabled', () => {
    const onChange = jest.fn();
    render(
      <Checkbox.Group disabled onChange={onChange}>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    const checkbox = screen.getByLabelText('Option 1');
    fireEvent.click(checkbox);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should not call onChange when readonly', () => {
    const onChange = jest.fn();
    render(
      <Checkbox.Group readOnly onChange={onChange}>
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    const checkbox = screen.getByLabelText('Option 1');
    fireEvent.click(checkbox);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should support vertical direction', () => {
    const { container } = render(
      <Checkbox.Group direction="vertical">
        <Checkbox value="1">Option 1</Checkbox>
        <Checkbox value="2">Option 2</Checkbox>
      </Checkbox.Group>,
    );
    expect(container.firstChild).toHaveClass('ice-checkbox-group-vertical');
  });
});
