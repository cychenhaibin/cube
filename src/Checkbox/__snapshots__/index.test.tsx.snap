// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<label
  class="ice-checkbox"
>
  <span
    class="ice-checkbox-wrapper"
  >
    <input
      class="ice-checkbox-input"
      type="checkbox"
      value=""
    />
    <span
      class="ice-checkbox-inner"
    />
  </span>
  <span
    class="ice-checkbox-label"
  >
    Checkbox
  </span>
</label>
`;

exports[`Checkbox.Group renders correctly 1`] = `
<div
  class="ice-checkbox-group ice-checkbox-group-horizontal"
  role="group"
>
  <label
    class="ice-checkbox"
  >
    <span
      class="ice-checkbox-wrapper"
    >
      <input
        class="ice-checkbox-input"
        type="checkbox"
        value="1"
      />
      <span
        class="ice-checkbox-inner"
      />
    </span>
    <span
      class="ice-checkbox-label"
    >
      Option 1
    </span>
  </label>
  <label
    class="ice-checkbox"
  >
    <span
      class="ice-checkbox-wrapper"
    >
      <input
        class="ice-checkbox-input"
        type="checkbox"
        value="2"
      />
      <span
        class="ice-checkbox-inner"
      />
    </span>
    <span
      class="ice-checkbox-label"
    >
      Option 2
    </span>
  </label>
</div>
`;
